File: cpp/src/guidance.cpp
#include "orchard/guidance.hpp"
#include <algorithm>
#include <cmath>
namespace orchard {
SteerCommand compute_steer(const RowPerception& perc, const Config& cfg) {
    SteerCommand cmd;
    cmd.lateral_error_m = perc.lateral_error_m;
    if (perc.heading_error_rad) cmd.heading_error_deg = *perc.heading_error_rad * 57.2957795f;
    cmd.confidence = perc.confidence;
    if (!perc.lateral_error_m && !perc.heading_error_rad) { cmd.hint = "HOLD"; return cmd; }
    const float lat_term = perc.lateral_error_m ? cfg.kp_lateral * *perc.lateral_error_m : 0.f;
    const float hdg_term = perc.heading_error_rad ? cfg.kp_heading * *perc.heading_error_rad : 0.f;
    cmd.steer = std::clamp(lat_term + hdg_term, -cfg.max_steer, cfg.max_steer);
    const bool lat_ok = perc.lateral_error_m && std::abs(*perc.lateral_error_m) <= cfg.deadband_m;
    const bool hdg_ok = cmd.heading_error_deg && std::abs(*cmd.heading_error_deg) <= cfg.deadband_deg;
    if (lat_ok && hdg_ok) { cmd.hint = "CENTER"; cmd.steer = 0.f; }
    else if (cmd.steer > 0.04f) cmd.hint = "LEFT";
    else if (cmd.steer < -0.04f) cmd.hint = "RIGHT";
    else cmd.hint = "CENTER";
    if (perc.lateral_error_m) {
        const float n = *perc.lateral_error_m / std::max(cfg.lightbar_full_scale_m, 1e-3f);
        cmd.lightbar = static_cast<int>(std::clamp(std::round(n * 5.f), -5.f, 5.f));
    }
    return cmd;
}
}

File: orchard_guidance/guidance.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

import numpy as np

from .config import GuidanceConfig
from .types import RowPerception


@dataclass
class SteerCommand:
    lateral_error_m: Optional[float]
    heading_error_deg: Optional[float]
    steer: float
    hint: str
    lightbar: int
    lightbar_cells: int = 11
    confidence: float = 0.0
    reason: str = ""


def compute_steer(perception: RowPerception, config: GuidanceConfig) -> SteerCommand:
    lat = perception.lateral_error_m
    hdg = None if perception.heading_error_rad is None else float(np.degrees(perception.heading_error_rad))
    if lat is None and hdg is None:
        return SteerCommand(None, None, 0.0, "HOLD", 0, confidence=0.0, reason="no_row")

    lat_term = 0.0 if lat is None else config.kp_lateral * lat
    hdg_term = 0.0 if hdg is None else config.kp_heading * np.deg2rad(hdg)
    steer = float(np.clip(lat_term + hdg_term, -config.max_steer, config.max_steer))
    lat_ok = lat is not None and abs(lat) <= config.deadband_m
    hdg_ok = hdg is not None and abs(hdg) <= config.deadband_deg
    if lat_ok and hdg_ok:
        hint, steer = "CENTER", 0.0
    elif steer > 0.04:
        hint = "LEFT"
    elif steer < -0.04:
        hint = "RIGHT"
    else:
        hint = "CENTER"
    half = 5
    bar = 0 if lat is None else int(np.clip(np.round((lat / max(config.lightbar_full_scale_m, 1e-3)) * half), -half, half))
    return SteerCommand(
        lat, hdg, steer, hint, bar, confidence=perception.confidence,
        reason="row_lock" if perception.left_line and perception.right_line else "partial",
    )
